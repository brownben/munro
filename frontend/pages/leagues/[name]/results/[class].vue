<script setup lang="ts">
import type { LeagueResultsOverview } from '~/api-types'
import type { Filters } from '~/utils/filter'

const route = useRoute()

const { data } = await useData<LeagueResultsOverview>(
  `leagues/${route.params.name}/results/${route.params.class}`,
)
const results = computed(
  () => data.value?.results.map(leagueResultWithAgeGender) ?? [],
)
const otherClasses = computed(
  () =>
    data.value?.classes.filter((course) => course !== data.value?.class_name) ??
    [],
)

const filters = reactive<Filters>({
  name: queryToString(route.query.name ?? ''),
  club: queryToString(route.query.club ?? ''),
  minAge: queryToString(route.query.minAge ?? '0'),
  maxAge: queryToString(route.query.maxAge ?? '100'),
  male: queryToString(route.query.male) !== 'false',
  female: queryToString(route.query.female) !== 'false',
})

if (data.value) {
  useTitle({
    title: `${data.value.class_name} Results - ${data.value.league}`,
    description: `League standings for the ${data.value.class_name} class in the ${data.value.league} league`,
  })
}
</script>
<template>
  <div v-if="data" class="flex h-full flex-col">
    <Heading
      :title="data.class_name"
      :link-location="`/leagues/${data.league}`"
      :link-text="data.league"
    >
      <template #rightAction>
        <ResultsFilterMenu v-model="filters" />
      </template>
    </Heading>

    <section
      v-if="results"
      class="mx-auto w-full flex-grow pb-10 pt-8 lg:px-8 print:sm:py-4"
      :class="{
        'max-w-screen-lg': results[0]?.points.length <= 3,
        'max-w-screen-xl':
          results[0]?.points.length > 3 && results[0]?.points.length <= 10,
      }"
    >
      <TableLeagueResults
        :results="results"
        :filters="filters"
        :events="data.events"
      />
    </section>
    <LinksSection
      v-if="otherClasses.length > 0"
      :links="
        otherClasses.map((course) => ({
          text: course,
          location: `/leagues/${data?.league}/results/${course}`,
        }))
      "
    >
      Other Classes
    </LinksSection>
  </div>
  <LazyError v-else />
</template>
