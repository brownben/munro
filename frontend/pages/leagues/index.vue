<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
