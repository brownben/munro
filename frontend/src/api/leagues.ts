import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
