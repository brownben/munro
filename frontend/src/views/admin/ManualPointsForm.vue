<template>
  <Layout title="Manual Points">
    <Meta title="Munro - Manual Points" description="" :block-robots="true" />
    <form class="col-span-2" @submit.prevent="addResult">
      <InputDropdown
        v-model="choices.league"
        :list="leagues.map((league) => league.name)"
        :include-blank="true"
        label="League:"
        url-parameter="league"
      />
      <InputDropdown
        v-model="choices.event"
        :list-with-different-value="
          eventsInLeague.map((event) => ({
            value: event.id,
            text: `${event.name} (${event.date})`,
          }))
        "
        :include-blank="true"
        label="Event:"
        class="mt-4"
        :validator="RequiredField('an event', true)"
      />
      <InputDropdown
        v-model="choices.course"
        :list="courses"
        :include-blank="true"
        label="Course"
        class="mt-4"
        :validator="RequiredField('a course', true)"
      />
      <InputDropdown
        v-model="choices.competitor"
        :list-with-different-value="
          competitorsInLeague.map((competitor) => ({
            value: competitor.id.toString(),
            text: competitorToText(competitor),
          }))
        "
        :include-blank="true"
        label="Competitor:"
        class="mt-4"
        :validator="RequiredField('a competitor', true)"
      />
      <InputNumber
        v-model.number="choices.points"
        label="Points:"
        class="mt-4"
        :min="0"
        :max="10000"
      />

      <button class="mt-8 button-lg">Add Result</button>
    </form>
  </Layout>
</template>

<script lang="ts" setup>
import { ref, computed } from 'vue'
import { useRouter } from 'vue-router'

import Layout from '../../components/Layout.vue'
import InputDropdown from '../../components/InputDropdown.vue'
import InputNumber from '../../components/InputNumber.vue'

import { RequiredField } from '../../scripts/inputValidation'

import { useMessages } from '../../store/messages'

import { useLeagues } from '../../api/leagues'
import { useEvents } from '../../api/events'
import { useCompetitors } from '../../api/competitors'
import { createManualResult } from '../../api/results'

const router = useRouter()
const messages = useMessages()

const [leagues] = useLeagues()
const [events] = useEvents()
const [competitors] = useCompetitors()
const choices = ref({
  league: '',
  event: '',
  course: '',
  competitor: '',
  points: 0,
})

const courses = computed(
  () =>
    leagues.value?.find((league) => league.name === choices.value.league)
      ?.courses ?? []
)
const eventsInLeague = computed(() =>
  events.value?.filter((event) => event.league === choices.value.league)
)
const competitorsInLeague = computed(() =>
  competitors.value
    ?.filter(
      (competitor) =>
        competitor.league === choices.value.league &&
        competitor.course === choices.value.course
    )
    ?.sort((a, b) => (a.name > b.name ? -1 : 1))
)

const competitorToText = (competitor: Competitor) => {
  if (competitor.club && competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}, ${competitor.club}) [${competitor.id}]`
  else if (competitor.club)
    return `${competitor.name} (${competitor.club}) [${competitor.id}]`
  else if (competitor.ageClass)
    return `${competitor.name} (${competitor.ageClass}) [${competitor.id}]`
  else return `${competitor.name} [${competitor.id}]`
}

const validateForm = () => {
  if (
    choices.value.event !== '' &&
    choices.value.course !== '' &&
    choices.value.competitor !== ''
  )
    return true
  else {
    messages.create(
      'Please Ensure Event, Course  and Competitor Fields are not Blank'
    )
    return false
  }
}

const addResult = () => {
  if (validateForm())
    createManualResult({
      competitor: Number(choices.value.competitor),
      event: choices.value.event,
      points: choices.value.points,
    })
      .then(() => router.push(`/leagues/${choices.value.league}/competitors`))
      .catch(() => false)
}
</script>
