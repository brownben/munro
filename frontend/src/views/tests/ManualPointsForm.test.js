import axios from 'axios'
import { mount, shallowMount } from '@vue/test-utils'
import ManualPointsForm from '@/views/ManualPointsForm'
import { sampleSingleLeague } from '@/tests/test data/leagues'
import { sampleThreeResultsDraw } from '@/tests/test data/results'

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = mount(ManualPointsForm, {
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = mount(ManualPointsForm, {
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Results Sorted Correctly', () => {
  const wrapper = mount(ManualPointsForm, {
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
  wrapper.setData({ course: '1', event: '', results: sampleThreeResultsDraw })
})

test('Return from Page', () => {
  const mockRouterPush = jest.fn()
  const mockAddMessage = jest.fn()
  const wrapper = mount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
      $router: { push: mockRouterPush },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.vm.returnToCompetitorsPage({ data: { message: 'Hello' } })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Hello')
  expect(mockRouterPush).toHaveBeenCalledTimes(1)
  expect(mockRouterPush).toHaveBeenLastCalledWith('/competitors')
})

test('Elapsed Time', () => {
  const wrapper = mount(ManualPointsForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.elapsedTime(0)).toBe('00:00')
  expect(wrapper.vm.elapsedTime(1)).toBe('00:01')
  expect(wrapper.vm.elapsedTime(45)).toBe('00:45')
  expect(wrapper.vm.elapsedTime(60)).toBe('01:00')
  expect(wrapper.vm.elapsedTime(83)).toBe('01:23')
  expect(wrapper.vm.elapsedTime(1023)).toBe('17:03')
  expect(wrapper.vm.elapsedTime(3600)).toBe('60:00')
  expect(wrapper.vm.elapsedTime(5117)).toBe('85:17')
})

test('Competitor Transform for Dropdown Input', () => {
  const wrapper = shallowMount(ManualPointsForm, {
    stubs: ['router-link', 'vue-headful'],
  })
  expect(
    wrapper.vm.competitorTransformForSelect({ name: 'Bob Jones', id: 22 })
  ).toEqual('Bob Jones [22]')
  expect(
    wrapper.vm.competitorTransformForSelect({
      name: 'Bob Jones',
      id: 22,
      ageClass: 'M45',
    })
  ).toEqual('Bob Jones (M45) [22]')
  expect(
    wrapper.vm.competitorTransformForSelect({
      name: 'Bob Jones',
      id: 22,
      club: 'HAT',
    })
  ).toEqual('Bob Jones (HAT) [22]')
  expect(
    wrapper.vm.competitorTransformForSelect({
      name: 'Bob Jones',
      id: 22,
      club: 'HAT',
      ageClass: 'M45',
    })
  ).toEqual('Bob Jones (M45, HAT) [22]')
})

test('Get Leagues - Request Called Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague })
  await wrapper.vm.getLeagues()
  expect(axios.get).toHaveBeenCalledWith('/api/leagues')
})

test('Get Leagues - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Leagues - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching Leagues'
  )
})

test('Get Competitors - Request Called Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague })
  await wrapper.vm.getCompetitors()
  expect(axios.get).toHaveBeenLastCalledWith('/api/competitors')
})

test('Get Competitors - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Competitors - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching Competitors'
  )
})

test('Courses in League', () => {
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: 'a', leagues: [] })
  expect(wrapper.vm.coursesInLeague).toEqual([])
  wrapper.setData({
    leagues: [
      { name: 'a', courses: [2] },
      { name: 'b', courses: 3 },
      { name: 'c', courses: [4] },
    ],
  })
  expect(wrapper.vm.coursesInLeague).toEqual([2])
  wrapper.setData({
    league: 'c',
    leagues: [
      { name: 'a', courses: 2 },
      { name: 'b', courses: 3 },
      { name: 'c', courses: [4] },
    ],
  })
  expect(wrapper.vm.coursesInLeague).toEqual([4])
})

test('Validate', () => {
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ event: '', competitor: '', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: 'a', competitor: '', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: '', competitor: 'a', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: '', competitor: '', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: 'a', competitor: 'a', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: 'a', competitor: '', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: '', competitor: 'a', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ event: 'a', competitor: 'a', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeTruthy()
})

test('Get Events - Request Called Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.resetAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleLeague })
  await wrapper.vm.getEvents()
  expect(axios.get).toHaveBeenLastCalledWith('/api/events')
})

test('Get Events - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Events - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching Events'
  )
})

test('Elapsed Time to Seconds', () => {
  const wrapper = shallowMount(ManualPointsForm, {
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.elapsedTimeToSeconds('00:00')).toBe(0)
  expect(wrapper.vm.elapsedTimeToSeconds('00:17')).toBe(17)
  expect(wrapper.vm.elapsedTimeToSeconds('00:30')).toBe(30)
  expect(wrapper.vm.elapsedTimeToSeconds('00:59')).toBe(59)
  expect(wrapper.vm.elapsedTimeToSeconds('01:00')).toBe(60)
  expect(wrapper.vm.elapsedTimeToSeconds('60:00')).toBe(3600)
  expect(wrapper.vm.elapsedTimeToSeconds('01:42')).toBe(102)
  expect(wrapper.vm.elapsedTimeToSeconds('37:37')).toBe(2257)
})

test('Competitors for League', () => {
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ course: '2', league: 'b', leagues: [] })
  expect(wrapper.vm.competitorsForLeague).toEqual([])
  wrapper.setData({
    competitors: [
      { name: 'a', course: '2', league: 'b' },
      { name: 'b', course: '2', league: 'b' },
      { name: 'c', course: '2', league: 'b' },
    ],
  })
  expect(wrapper.vm.competitorsForLeague).toEqual([
    { name: 'a', course: '2', league: 'b' },
    { name: 'b', course: '2', league: 'b' },
    { name: 'c', course: '2', league: 'b' },
  ])
  wrapper.setData({
    league: 'c',
    course: '2',
    competitors: [
      { name: 'a', course: '2', league: 'b' },
      { name: 'b', course: '2', league: 'b' },
      { name: 'c', course: '2', league: 'b' },
    ],
  })
  expect(wrapper.vm.competitorsForLeague).toEqual([])
  wrapper.setData({
    league: 'c',
    course: '2',
    competitors: [
      { name: 'q', course: '2', league: 'c' },
      { name: 'b', course: '2', league: 'b' },
      { name: 'p', course: '2', league: 'c' },
    ],
  })
  expect(wrapper.vm.competitorsForLeague).toEqual([
    { name: 'p', course: '2', league: 'c' },
    { name: 'q', course: '2', league: 'c' },
  ])
})

test('Add Result - Invalid', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.addResult()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Select a Competitor and an Event'
  )
})

test('Add Result', async () => {
  const mockAddMessageFunction = jest.fn()
  const mockReturnToCompetitorsPage = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setMethods({ returnToCompetitorsPage: mockReturnToCompetitorsPage })
  wrapper.setData({
    event: '1 - 2',
    course: '2',
    events: [{ id: '4', name: '1', date: '2' }],
    competitor: '1',
    points: 7,
  })
  axios.get.mockResolvedValue()
  axios.post.mockResolvedValue()
  await wrapper.vm.addResult()
  expect(axios.post).toHaveBeenCalled()
  expect(axios.post).toHaveBeenLastCalledWith('/api/results/manual', {
    competitor: '1',
    points: 7,
    event: '4',
  })
  expect(wrapper.vm.returnToCompetitorsPage).toHaveBeenCalled()
  expect(mockAddMessageFunction).not.toHaveBeenCalled()
})

test('Add Result - No Event', async () => {
  const mockAddMessageFunction = jest.fn()
  const mockReturnToCompetitorsPage = jest.fn()
  const wrapper = shallowMount(ManualPointsForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setMethods({ returnToCompetitorsPage: mockReturnToCompetitorsPage })
  wrapper.setData({
    event: '1 - 2',
    course: '2',
    events: [],
    competitor: '1',
    points: 7,
  })
  axios.get.mockResolvedValue()
  axios.post.mockResolvedValue()
  await wrapper.vm.addResult()
  expect(axios.post).not.toHaveBeenCalled()

  expect(wrapper.vm.returnToCompetitorsPage).not.toHaveBeenCalled()
  expect(mockAddMessageFunction).toHaveBeenCalled()
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('No Event Selected')
})
