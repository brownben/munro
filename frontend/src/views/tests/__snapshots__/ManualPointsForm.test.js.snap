// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div
  class="view"
>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Manual Points"
  />
   
  <h1
    class="mb-2 text-3xl font-normal text-main font-heading"
  >
    
    Manual Points
  
  </h1>
   
  <form>
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        League:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Event:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Course
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Competitor:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative appearance-none text-input top--4"
    >
      <label
        class="relative p-1 pb-0 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Points:
      </label>
       
      <input
        class="w-full px-4 py-2 border outline-none appearance-none border-main rounded-shape text-body"
        max="10000"
        min="0"
        type="number"
      />
    </div>
     
    <button
      class="button-lg"
    >
      Add Result
    </button>
  </form>
</div>
`;

exports[`Results Sorted Correctly 1`] = `
<div
  class="view"
>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Manual Points"
  />
   
  <h1
    class="mb-2 text-3xl font-normal text-main font-heading"
  >
    
    Manual Points
  
  </h1>
   
  <form>
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        League:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Event:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Course
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative dropdown-input top--4"
      include-blank="true"
    >
      <label
        class="relative p-1 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Competitor:
      </label>
       
      <div
        class="visible w-full px-4 py-2 border outline-none border-main rounded-shape"
      >
        <p
          class="h-6 leading-normal truncate"
        >
          
        </p>
         
        <svg
          class="float-right h-8 fill-current text-main mr--3 mt--68/10"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 7.84L12 12.42l4.59-4.58L18 9.25l-6 6-6-6z"
          />
           
          <path
            d="M0-.75h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
       
      <div
        class="absolute z-40 w-full mb-2 text-center bg-white border border-t-0 select-none dropdown text-body border-main"
        name="open"
        style="display: none;"
      />
    </div>
     
    <div
      class="relative appearance-none text-input top--4"
    >
      <label
        class="relative p-1 pb-0 mx-3 text-sm bg-white select-none font-heading text-main bottom--3"
      >
        Points:
      </label>
       
      <input
        class="w-full px-4 py-2 border outline-none appearance-none border-main rounded-shape text-body"
        max="10000"
        min="0"
        type="number"
      />
    </div>
     
    <button
      class="button-lg"
    >
      Add Result
    </button>
  </form>
</div>
`;
